File: Backtracking/BFS/103BinaryTreeZigzaLevelOrdergraversal.py
# Definition for a binary tree node.
import collections
from typing import List, Optional

# 经典BFS
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root: return []

        res = []
        q = collections.deque([])
        q.append(root)
        direction = -1
        while(q):
            level = []
            for i in range(len(q)):
                cur = q.popleft()
                if cur:
                    level.append(cur.val)
                    q.append(cur.left)
                    q.append(cur.right)
            if len(level) != 0: 
                if direction == 1:
                    level.reverse()
                direction = -1 * direction
                res.append(level)
        return res

        

File: Backtracking/BFS/Tree/103BinaryTreeZigzaLevelOrdergraversal.java
package Backtracking.BFS.Tree;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Queue;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        Queue<TreeNode> q = new ArrayDeque<>();
        if (root == null) {
            return res;
        }

        Boolean leftToRight = false;
        q.offer(root);
        while (!q.isEmpty()) {
            List<Integer> curLevel = new ArrayList<Integer>();
            int level = q.size();
            for (int i = 0; i < level; i++) {
                var cur = q.poll();
                curLevel.add(cur.val);
                if (cur.right != null) {
                    q.offer(cur.right);
                }
                if (cur.left != null) {
                    q.offer(cur.left);
                }
            }

            if (!leftToRight) {
                Collections.reverse(curLevel);
            }
            res.add(curLevel);
            leftToRight = !leftToRight;
        }
        return res;
    }
}